prover/src/prover.rs

```diff
@@ -1,170 +1,216 @@
-            proof: proof_detail.proof_data,
-            ..Default::default()
-        };
+impl LocalProver {
+    pub fn new(config: LocalProverConfig, prover_types: Vec<ProverType>) -> Self {
+        let circuits_handler_provider = CircuitsHandlerProvider::new(config.clone())
+            .context("failed to create circuits handler provider")
+            .unwrap();
 
-        self.do_submit(&request)
+        Self {
+            config,
+            prover_types,
+            circuits_handler_provider: RwLock::new(circuits_handler_provider),
+            next_task_id: Arc::new(Mutex::new(0)),
+            current_task: Arc::new(Mutex::new(None)),
+            // result: Arc::new(Mutex::new(Err(anyhow::Error::msg("prover not started")))),
+        }
     }
 
-    pub fn submit_error(
+    async fn do_prove(
         &self,
-        task: &Task,
-        failure_type: ProofFailureType,
-        error: Error,
-    ) -> Result<()> {
-        log::info!("[prover] start to submit_error, task id: {}", task.id);
-        let request = SubmitProofRequest {
-            uuid: task.uuid.clone(),
-            task_id: task.id.clone(),
-            task_type: task.task_type,
-            status: ProofStatus::Error,
-            failure_type: Some(failure_type),
-            failure_msg: Some(format!("{:#}", error)),
-            ..Default::default()
+        req: ProveRequest,
+        handler: Arc<Box<dyn CircuitsHandler>>,
+    ) -> Result<ProveResponse> {
+        let task_id = {
+            let mut next_task_id = self.next_task_id.lock().unwrap();
+            *next_task_id += 1;
+            *next_task_id
         };
-        self.do_submit(&request)
+
+        let duration = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let created_at = duration.as_secs() as f64 + duration.subsec_nanos() as f64 * 1e-9;
+
+        let req_clone = req.clone();
+        let handle = Handle::current();
+        let task_handle =
+            tokio::task::spawn_blocking(move || handle.block_on(handler.get_proof_data(req_clone)));
+
+        *self.current_task.lock().unwrap() = Some(task_handle);
+
+        Ok(ProveResponse {
+            task_id: task_id.to_string(),
+            circuit_type: req.circuit_type,
+            circuit_version: req.circuit_version.clone(),
+            hard_fork_name: req.hard_fork_name.clone(),
+            status: TaskStatus::Proving,
+            created_at,
+            started_at: None,
+            finished_at: None,
+            compute_time_sec: None,
+            input: Some(req.input),
+            proof: None,
+            vk: None,
+            error: None,
+        })
     }
+}
 
-    fn do_submit(&self, request: &SubmitProofRequest) -> Result<()> {
-        self.coordinator_client.borrow_mut().submit_proof(request)?;
-        Ok(())
+fn build_prove_error_response(
+    task_id: String,
+    status: TaskStatus,
+    proof: Option<String>,
+    error_msg: String,
+) -> ProveResponse {
```

Comment:
Why don't add `Default` to ProveResponse, and only fill the specified filed, other use default instead